### ISSUE TITLE
State transfer for PBFT

### ISSUE THREAD

**Issue by kchristidis** [hyperledger-archives/fabric#291]
From [the paper](http://research.microsoft.com/en-us/um/people/mcastro/publications/p398-castro-bft-tocs.pdf), pg. 413 on view change:

> The primary updates its state to reflect the information in the `NEW-VIEW` message. It obtains any requests in `X` that it is missing and if it does not have the checkpoint with sequence number `h`, **it also initiates the protocol to fetch the missing state (see Section 6.2.2)**. When it has all requests in `X` and the checkpoint with sequence number h is stable, it records in its log that the requests are pre-prepared in view `v+1`.

We want to work on the sentence in **bold**.

Let's first figure out how #145 ties into this (@muralisrini), and what hooks are there (or we should expect?) from @srderson and @manish-sethi.

**Comment by srderson** [comment]
I opened issue #299 before seeing this issue. Let's meet to discuss.

**Comment by kchristidis** [comment]
@srderson: I left a comment.

@ghaskins: Keep a look at the discussion there as well, in case you do end up tackling this issue. In general, we should be simplifying the state transfer mechanism greatly compared to what's described in the paper. A lot of the low-level details of the paper either don't concern us at all, or have been covered by the work that @manish-sethi and @srderson have done. The key thing from Section 6.2.2 is to identify the flow that's in place.

**Comment by kchristidis** [comment]
@jyellick please have a look at this as well. Post questions here. (Bug @vukolic, @corecode, and me for the protocol side of things, and @srderson, @manish-sethi for anything that has to do with getting data from the ledger or the DB.) Also see the discussion in #299.

**Comment by corecode** [comment]
See [PBFT TOCS](http://research.microsoft.com/en-us/um/people/mcastro/publications/p398-castro-bft-tocs.pdf) _Fig. 24_ `RETRANSMIT`.  The `STATE` message corresponds to both checkpoint and stack state.  The next line specifies retransmission of locally-known requests that match the digests in the `Xset`.

**Comment by jyellick** [comment]
A few questions.

In `pbft.go`, the `executeOne` function, the `stateHash` is presently hardcoded, with a comment referencing that the checkpoint should be obtained from CPI, and then the state hash should be created.  Do we have a plan for this?  It looks like the checkpoint hash is the same as the `stateHash` for the corresponding last block written, and in some ways this would seem to make life easy, but it seems to have a couple potential problems.

When null requests are committed, the blockchain is not affected, and therefore would not affect the `stateHash` above.  Is this a problem?  In particular, it seems like the hash would not distinguish between these two sequences:

a) <1, Block A>, <2, null>, <3, Block B>
b) <1, null>, <2, Block A>, <3, Block B>

My guess is that scenario probably wouldn't matter, it is the integrity of the blockchain that we care about, and we discard the sequence numbers and null requests after checkpointing anyhow, but it is a deviation from the paper.

This brings me to my second question, how do we map sequence numbers to blocks? Assume a peer is at sequence number 40 and block number 35 and receives a stable checkpoint for sequence number 50.  The peer needs to retrieve the missing blocks through the checkpoint, but it does not know how many blocks that is.  The peer could pull blocks and compute hashes until finding the block matching the stable checkpoint, but this seems expensive, and possibly a way a malicious node could make state transfer slow.

Assuming we can map sequence numbers to block numbers, this is how I would envision a sample flow

1) Peer has transactions through checkpoint sequence number 40
2) Peer receives stable checkpoint for sequence number 50, initiates state transfer
3) Peer requests blocks `seqToBlockNumber(41)` through `seqToBlockNumber(50)` from any peer or peers
4) Peer validates `stateHash` of block `seqToBlockNumber(50)` against stable checkpoint
5) Peer validates blockchain from block `seqToBlockNumber(40)` through `seqToBlockNumber(50)`
6) Peer re-requests from a different peer any blocks which did not validate
7) Repeat 4-6 until all blocks are valid

This solves the problem of catching up to the checkpoint, as described in the paper.   It does not necessarily handle catching up to the current request if new requests are made during the state transfer.  The paper dismisses this as unlikely to be a problem in most scenarios, but as @srderson suggested in Issue #299, blockchain gives us the ability to instead catch up to the current state.   By modifying the process above to include fetching f+1 copies of the latest block instead of validating against the stable checkpoint, the peer should be able to catch all the way up (or at least be closer to caught up than via the checkpoint).  This seems to raise the problem though of the reverse mapping, of blocks to sequence numbers, though I haven't thought through this yet.

Finally, I just wanted to verify the intended entrypoint for the state transfer work.  In `pbft.go`, the `recvCheckpoint` function has a TODO which notes the low water mark will be set to 0 if the node does not have the quorum checkpoint.  This seems like an equivalent condition to the stable checkpoint being for a sequence number higher than the low water mark in the log, so I think this is where we should trigger a state transfer?

**Comment by corecode** [comment]
How about we add the block number to the PBFT `chkpts` value?  Then we know exactly which block to fetch, and how to verify it (state).

This still requires a way to sync a state which corresponds to a checkpoint, which will have to be exposed by the ledger code (see #299).

There are several entry points for state transfer.  I've marked them more clearly in my devel branch.  Triggers for state transfer are:
- initial startup
- (maybe) receiving a stable checkpoint which we do not have yet
- processing of new-view message failing due to missing state/requests.

**Comment by corecode** [comment]
Maybe it makes sense to first add a way to synchronize requests when checkpoints are up to date.  This is required, e.g. for recovery after byzantine primary (see `TestInconsistentDataViewChange`)

**Comment by kchristidis** [comment]
@jyellick a few follow-up questions:
1. "...but it is a deviation from the paper." Can you explain how is that?
2. "It does not necessarily handle catching up to the current request if new requests are made during the state transfer." Can't we be listening to incoming requests and buffer while we're restoring our state, so that we execute them right after and catch up? (Assuming the rate execution is higher than the rate on request reception.)

**Comment by jyellick** [comment]
@kchristidis 
1. The key deviation from the paper was that null requests would not affect the state hash.  Thanks to issue #349 this should no longer be the case.  Issuing a null request will change the sequence number embedded in the block for future transactions, which will affect the block state hash, so I think this is solved.
2.  Yes, the state transfer should be able to execute concurrently as incoming requests arrive and are buffered.  This works perfectly for 2 of 3 cases (view change, and witnessed higher checkpoint), but has some caveats for the initial startup case.

View Change: In the 'good path', the replica has copies of all requests since the last checkpoint, so syncing to the most recent checkpoint, and given ordering from the new primary, the replica can execute the requests to get to the current state.

Witnessed Higher Checkpoint: In this case, because a higher checkpoint was just witnessed, there should be no sequence numbers assigned higher than this checkpoint, the replica should witness new requests, and can execute normally.

Initial startup:  This is the tricky case.  In the paper, the replica broadcasts a message requesting information for the partition with sequence number -1, and receives a weak certificate indicating the most recent checkpoint number, then proceeds with the state replication in the normal path.  The problem here, is although the replica can buffer new requests, it may have already missed some requests which occurred before startup.  In this case, it will never catch up until ~~either a view change,~~ (a view change contains the request digests, but not the requests themselves) or a witnessed higher checkpoint occur.  If requests are frequent, a new checkpoint should be reached quickly, but without extending the notion of state replication beyond checkpoints, this seems unavoidable.

I would suggest that for MVP, and perhaps longer term, the caveat on initial startup is acceptable.  If the scenario described for initial startup is a problem, I would propose the following solution.   Remove the initial startup case from state replication.  Instead, at initial startup, the replica should broadcast a request for the current blockchain state, receiving a weak certificate from f+1 other replicas, and then sync its blockchain to the others.  Consensus can then start with the sequence number set to 1 + the sequence number from the most recent block.  If requests have been arriving rapidly, then this state may be stale, and the replica will need to wait for a new checkpoint (but this should be quick, as requests are arriving rapidly).  If requests are infrequent, then the replica will be up to date and ready to process requests.

**Comment by kchristidis** [comment]
@jyellick: As we discussed on the phone just now, a null request won't affect the global state hash (which is different from the block state hash), so what you wrote on [your earlier post](https://github.com/openblockchain/obc-peer/issues/291#issuecomment-161453758) stands. This may not be a problem, in the sense that it can't be exploited in a Byzantine manner.

For initial startup, the idea of waiting for two stable checkpoints sounds appealing to me. We can discuss both with @vukolic on Monday to make sure.

**Comment by jyellick** [comment]
@kchristidis: Thanks for clearing that up, since the global state hash will not incorporate the sequence number, we still have some work to do.  There are two paths I see here:
1. Concatenate the sequence number and the global state hash and use this as the checkpoint value.
2. Instead of using the state hash, use the block hash.  The block contains the sequence number, and the state hash, so, this incorporates both without performing additional computations as well.

I'd like for @vukolic and perhaps others more familiar with the blockchain side to weigh in here, but the block hash seems like a better choice for a number of reasons.  In order to catch up on missed transactions, we'll already need to retrieve all missing blocks.  Additionally, my understanding is that the state hash will by design assume the same value for the same states.  When paired with a sequence number this is probably fine, but still makes me a little uncomfortable.

Secondly, as has been noted by others, I think it's worth looking at exactly how PBFT is doing checkpointing in the context of blockchain.  The purpose of checkpointing infrequently seems primarily to be to reduce the number of hash computations performed, but the blockchain is performing the hash computations with every transaction anyhow.  Would there be any significant drawback to checkpointing after every transaction?  This has the added benefit of simplifying many of the cases described above.  If every transaction is a checkpoint, the most recent checkpoint is always the last transaction, so there is no possibility to be up to checkpoint without being completely up to date.

**Comment by jyellick** [comment]
@vukolic In particular, but anyone else with insight is welcome to join in.  I'm trying to reconcile some statements from the paper.  

In section 4.4 the paper states

> The checkpoint protocol is used to advance the low and high water marks (which limit what messages will be added to the log). The low water mark h is equal to the sequence number of the last stable checkpoint and the high water mark is H = h + L, where L is the log size. The log size is the maximum number of consecutive sequence numbers for which the replica will log information. It is obtained by multiplying K by a small constant factor (e.g., 2) that is big enough so that it is unlikely for replicas to stall waiting for a checkpoint to become stable.

And later, in section 6.2.2 it states

> A replica initiates a state transfer when it learns about a stable checkpoint with sequence number greater than the high water mark in its log. It uses the state transfer mechanism to fetch modifications to the service state that it is missing. The replica may learn about such a checkpoint by receiving CHECKPOINT messages or as the result of a view change.

But how could we learn about a checkpoint beyond our high watermark?  We discard messages if their sequence number is outside of the water marks, and this includes checkpoint messages.  (This is confirmed by the psuedo-code for `RECEIVE(<CHECKPOINT,...>)`  in Fig 23.)  This makes sense because tracking checkpoint messages outside our water marks could cause our memory usage to grow in an unbounded way, but for a node which falls behind substantially (more than L), then we'll never observe another stable checkpoint.

Maybe there is an easy way around this, maybe we keep only the checkpoint message with the highest sequence number from each replica, and claim that 2f+1 checkpoint messages past a particular checkpoint indicate that checkpoint must have been stable at some point.  Still, the correctness of such a scheme is not immediately obvious.

**Comment by vukolic** [comment]
@jyellick What you say makes sense. One should not discard all checkpoint msgs higher than the high watermark, nor keep them all around. But keeping one such checkpoint msg (higher than the high watermark) per replica would make sense. 

Strictly speaking 2f+1 checkpoint messages past a particular checkpoint still does not guarantee a _stable_ checkpoint at some other replica, but having f+1 checkpoint msgs past a particular checkpoint means that at least one correct replica has reached the f+1st highest checkpoint no. (which may not be stable at that replica).

Hope this helps, if not - let me know.

**Comment by vukolic** [comment]
BTW, whoever is working on checkpointing, state transfer and persisting to non-volatile storage, I strongly suggest reading the Usenix ATC'13 paper by Bessani et al, (http://www.di.fc.ul.pt/~bessani/publications/usenix13-dsmr.pdf)

**Comment by kchristidis** [comment]
There's still work going on related to this issue (see [here](https://github.com/openblockchain/obc-peer/issues/299#issuecomment-163384508) for example) which pushes back the delivery of this. Moving it to Sprint 4. @jyellick feel free to update with more details.

**Comment by jyellick** [comment]
# State Transfer Design

@kchristidis and @vukolic in particular, but as always, I would welcome comments from others, especially @srderson to make sure the blockchain operations are correct.

## Differences in State Transfer from the Paper

The paper describes a tree for meta-data and data storage.  (See figure 6.)  Trees corresponding to stable checkpoints are garbage collected when a new checkpoint becomes stable.  This introduces complexity which does not need to be addressed when using blockchain as the backing data structure.

The blockchain has the property, that given a correct copy of the latest block, it is possible to verify all other blocks, and given a correct state for any block, it is possible to play the transactions forwards (via state deltas, or applying transactions) to the state for any later block, validating the state against the blockchain at each step.  Therefore, if we assign the hash of the corresponding block to be the checkpoint value, then it is possible for a replica to reply with verifiable data which can be played forward to the next stable checkpoint.

Because of this, it is no longer necessary to broadcast a `FETCH` to all replicas.  This simplifies the transfer process substantially.

## Limitations of a State Transfer

Replicas are only capable of validating blocks which correspond to a checkpoint.  Although it is intuitive that a replica could request the current state, and f+1 matching copies of the corresponding block to validate it, this actually creates more problems that it solves.  Replicas are only guaranteed to have a block with sequence number n after they have issued a `CHECKPOINT` with a sequence number great than n.  Therefore, attempting to fetch f+1 matching copies of the block would only be guaranteed to succeed after the `CHECKPOINT` message which includes the block hash of its corresponding (and more recent) block.

Therefore, although we fetch the current state, we do not attempt to validate that state (nor can we) until the next `CHECKPOINT` occurs.  In the future, when state deltas can be applied to play the state backwards, it would be possible to validate the state from f+1 blocks at the last checkpoint, but this would only potentially reduce the time to sync, correctness is possible without it.

So, in general, state transfers must occur in a two stage process before a replica to begin participating again normally.  
1. Become aware of a checkpoint for which the replica can never receive a quorum certificate
   - If a replica can never receive a quorum certificate for a checkpoint, then it can never advance its water marks, and it will stop executing requests
   - This usually implies the replica has already missed requests beyond that checkpoint
   - Therefore initiating a state transfer to that checkpoint cannot catch the node up completely
   - This does however allow the replica to move its watermarks and begin logging requests for when it does catch up
   - This triggers the start of a state transfer
2. Witness a new stable checkpoint
   - The replica now has a hash it can validate the blockchain against
   - Using the validated blocks, the replica can validate the state
   - The replica can either replay a small number of transactions, or apply a small number of state deltas to catch up (see discussion at bottom)
   - The replica can then finish state transfer

### When a Replica Has Fallen Behind

This scenario could occur if there is a network outage, the system temporarily becomes resource constrained and cannot execute transactions fast enough, the replica process is killed, or any other fault which prevents its sequence numbers from advancing.
1. Become aware of a stable checkpoint which the replica can never reach
   1. The replica listens for `CHECKPOINT` messages, and records the last checkpoint sequence number observed for each replica if it exceeds the high water mark. (This prevents the unbounded growth of tracking messages outside of the watermarks)
   2. If the checkpoint message is higher than the high water mark then sort the checkpoint sequence numbers, and discard the highest f of them
   3. If the maximum checkpoint sequence number (of the remaining) is above the high water mark, then this replica will never have a quorum certificate for that checkpoint. (There are a total of 3f+1 possible checkpoint messages for that sequence number, this replica will not generate one (because it has missed messages), and it has missed f checkpoint messages alreeady, therefore there are only 2f `CHECKPOINT` messages it can possibly observe)
   4. The replica sets the low water mark to the missed checkpoint sequence number and marks itself out of date
   5. The replica initiates the state transfer (indicating the replica which sent the `CHECKPOINT` message as the source)
2. Witness a new stable checkpoint
   1. The state transfer can now complete
   2. The replica is marked as up to date, and begins executing the requests that it has been queuing in its log

### At startup

For now, it is assumed that the PBFT network is up and functioning under normal conditions of no more than f faulty nodes (including this starting replica).  Starting PBFT when all replicas are stopped requires that f+1 nodes will attest to the same current state and can be considered a separate path from state transfer.

Presently, the `HelloMessage` contains only the latest block number.  To make startup recovery more efficient, it would be nice to enhance this message to include the data referenced by its latest block's `consensusMetadata`.  However, without this information, the replica can actually proceed exactly as in the 'Replica Has Fallen Behind' scenario, and should be caught up within two checkpoints.

The replica will be out of sync until at least 2 stable checkpoint messages have occurred.

### At View Change

This proceeds exactly as with 'Replica Has Fallen Behind', except that the unreachable stable checkpoint is discovered via the new view. 

## Performing the State Transfer
1. Request the current state from some peer (specified by the requester) 
2. If the state returned is not after the last checkpoint, repeat 1 with some other peer
3. Wait for the next checkpoint quorum certificate
4. Request the blockchain from the block referred to by the checkpoint, through the block referred to by the state retrieved in 1, from some peer in the checkpoint quorum
5. If the blockchain retrieved does not validate against the checkpoint, or against itself (using the `Block.previousBlockHash`), then repeat from 3 with a different peer for any nonvalidating blocks (The validation will actually be performed by the utility from issue #367)
6. Validate the state hash against the `Block.stateHash`, if this fails repeat from 1 with some other peer
7. The replica plays the state forward, using either state deltas or executing transactions (see below), verifying the state hash at each step
8. The replica marks itself as up to date, and begins executing transactions again normally

### Playing Transactions vs Applying State Deltas

Depending on the complexity of transactions vs. the size of state deltas, it's not always clear which will be faster.

@binhn  has expressed a strong preference towards playing the state deltas forward, and for many batched transactions, this seems likely to be faster.  However, this creates some potential problems.  Primarily, receiving the state could take time, and blocks/transactions are continually processed.  If the block creation rate is fast enough, it is possible that the state deltas may have been garbage collected and no longer available.  The paper addresses this by broadcasting the message for the content, and in the case that no one has that content available anymore (receiving replies indicating such), then replica syncs to a newer state.  This creates the potential the the replica might never catch up.

My suspicion, is that if the state deltas are being discarded faster than we can retrieve them, then this indicates the transactions are very fast to execute, and that rather than attempting to pull a new state, we should simply play the transactions forward.

Therefore I would propose that on failure to retrieve a state delta, this triggers the transaction execution path.

**Comment by jyellick** [comment]
A quick update on the status of this item.
- Code for the function is complete.
- Test coverage of the normal path operation is complete, and passing.
- Test coverage for the byzantine path is partially complete
- Integration into the main pbft path is partially complete
- Testing is currently being done using a mock ledger implementation, pending integration of the state delta transfer mechanism

Work is being done in the https://github.com/jyellick/obc-peer/tree/jyellick-pbft-state-transfer branch if anyone wants to take a look prior to completion.

**Comment by binhn** [comment]
@jyellick I like your analysis and the approach. I do have 1 question: If we can't get the delta and we gets error during exec txs or the resulting hash doesn't match, what path do we take? At some point, we might have to give up the state transfer, and so we need to handle it gracefully.

**Comment by corecode** [comment]
I don't know how this relates exactly, but sieve (filtering of non-deterministic transactions) never passes non-deterministic requests into the PBFT core.  @binhn does this address your concern?

**Comment by srderson** [comment]
@jyellick PR #418 contains API for rolling back an applied delta as discussed in issue #414. It currently allows one delta apply followed by a commit or rollback. Would you ever need to rollback a batch of deltas?

Also, the term rollback is now confusing because it means two things.
1) In this case, I'm talking about removing a malicious delta after checking the hash.
2) Other case using a trusted delta to roll the state forwards or backwards.
Maybe we need another term to reduce confusion? Let me know if you have any suggestions.

**Comment by jyellick** [comment]
@srderson : I think the single state delta rewind works for this issue.  For the SyncStateDeltas message, there may be multiple state deltas contained, but they should each be able to play the state to a statehash contained in a block.  Because of this, we should only ever need to rewind a single applied state delta.  The only case where multiple deltas need to be applied prior to checking that hash is when a full snapshot is being synced.  There we would simply empty the state, so rollback is unnecessary.  

In regards to the term rollback, it does seem appropriate for this case, but I understand the possible confusion it could create.  If I had to pick a synonym, 'Unapply' jumps to mind.

@binhn For scenarios such as you describe, the default recovery is to discard the current state, request a new state from some other replica, and attempt to play it to a new checkpoint.  If f+1 nodes cannot reply with appropriate state deltas, either the byzantine failure conditions have been violated, or we have fallen behind again, either way, we will continue to participate in transaction ordering, but not execute any transactions until we are back in a good state.

For now, the code only relies on playing the state deltas forward, and if transactions are executing faster than the deltas can be recovered, then we will simply lag behind until the system slows enough for us to catch up.  Including the option to switch to transaction execution as described is something that would be nice to include in the longer term though.

**Comment by binhn** [comment]
@corecode @jyellick thanks for the update. We hope sieve can detect but would not completely eliminate ND txs from entering the blockchain. It will definitely complement other techniques, but in exec txs, we have to assume the possibility of ND, even minute possibility. So chances for ending up with a different hash exists, and we have to handle. Getting the deltas randomly from f+1 peers would help overcome as described.

@muralisrini We also need some indicators to properly respond to state query when the VP is in transition and can't exec tx

**Comment by corecode** [comment]
Sieve is designed to handle this.  If a quorum of f+1 or more replicas report the same result, the transaction is accepted as deterministic (enough).  However, in the worst case, this could mean that f out of these f+1 are byzantine, and only one correct replica executed the transaction with the accepted result.

The sync code for sieve will have to make sure to sync from these f+1 replicas.  However, sieve is synchronous and does not require re-execution of transactions.  Any time a replica falls behind (or encountered a non-matching result of a transaction that was accepted as "deterministic enough"), it will have to sync the quorum accepted result (no re-execution).

**Comment by vukolic** [comment]
Expanding on what @corecode wrote above. We will have two consensus protocols PBFT and Sieve. PBFT is the classical one and non-deterministic transactions are ruled out by assumption (in practice - other layers of code need to handle non-determinism). Sieve, on the other hand, is built on PBFT and deals with non-determinism.

In Sieve, replicas agree on a single output hash _h_, or quarantine (i.e., _sieve out_) the transaction if non-determinism is detected. Agreement is possible only if f+1 replicas reach the same output hash _h_. Given a transaction/block number - we denote this set of replicas as _dSet_ (decision set).

An unavoidable problem in detecting non-deterministic tx is that false negatives are possible, i.e., a transaction might be non-deterministic but can pass Sieve (or any similar protocol) as deterministic (for example, imagine 3 replicas execute code that rolls a dice and all 3 replicas roll number 6 - roll of the dice is obviously non-deterministic but it appears to replicas as deterministic). In that case, a replica catching up with state transfer (or the remaining 4th replica that - in our example - rolled number 4) might need to fetch the state that matches agreed hash _h_ from one of the f+1 replicas belonging to _dSet_ (i.e,. those replicas that initially speculatively executed the transaction and reached output state that hashes to _h_).

**Comment by corecode** [comment]
@jyellick the whole syncing seems quite independent of PBFT, and very specific to OBC.  Maybe it should live in consensus, or consensus/helper, so that noops can use it as well?

**Comment by jyellick** [comment]
@corecode I'd agree there's definitely some common code that could probably be factored out.  In particular, the blockchain replication beyond the block needed to validate the current state is totally superfluous from a consensus perspective, though necessary from an OBC one.  I'd be open to suggestions for where to refactor things to.  Presently, I'm working on integrating this work with your work as well as the work of @kchristidis , and I think it makes sense to get the current structure working in place before attempting to refactor.  

On a somewhat related note, I discussed briefly with @kchristidis about using state transfer for sieve.  I think for sieve, state transfer in its totality is overkill.  The assumption for state transfer as designed here is that the current state is actually corrupt (or very far behind), and there is no way to easily transform the replica's state into the correct current state.  For sieve, my understanding is that after receiving a weak certificate for the result of a transaction, if your result differs, then you need only to rollback to before the transaction, retrieve a single state delta, and verify that the result matches the transaction hash in that weak certificate.  In this case, I'd suggest you access the ledger API directly, and if some failure occurs trigger the full state transfer.  The full state might be gigabytes in size, and transferring and re-hashing it could be a substantial cost.  Rolling back and playing deltas seems like the superior option.

**Comment by jyellick** [comment]
@corecode @kchristidis As of commit https://github.com/jyellick/obc-peer/commit/562d3e8c2c097050fe0b9487f7d56f8d7a7a8826 all pbft tests are passing, and the global tests are capable of running, though many fail (this appears to be the case in kostas/devel as well)

The state transfer and pbft network tests are still largely distinct, but compatible, further integration and mock network testing for the state transfer is an outstanding TODO.

There appears to be one missing backing API pending from issue #299 , but integration should be fairly easy after its delivery.

**Comment by jyellick** [comment]
As of commit https://github.com/jyellick/obc-peer/commit/d88028d12541b09117b5a6df674b2b7fdbe0c1d8 

Updates:
- The general state transfer mechanism has been split out into its own package, `statetransfer`
- The state transfer API to be exposed to other consensus packages is all implemented, including the synchronous state transfer mechanism requested by @corecode .  This is available via `SynchronousStateTransfer(blockNumber uint64, blockHash []byte, replicaIds []uint64) error`

Known Limitations:
- The tests remain in `obcpbft` until we can find a better way to share the test mock ledger implementation across packages.
- The config is still in pbft, this should probably be moved into state transfer
- Further unhappy path test cases need to be developed
- Completing the integration with the rest of OBC is still pending the completion of issue #299

**Comment by kchristidis** [comment]
Great, thanks for the update @jyellick :+1:

**Comment by jyellick** [comment]
Just a quick update.  As of commit 76f6e40b849ab3cbc76ce4801f8638f65b6198c1 , we now have a (mock network and mock ledger) based test passing which includes a replica falling behind, initiating state transfer, completing state transfer, and resuming normal operation in the PBFT network.

There is some remaining integration work pending the completion of #299 , and as mentioned before, additional testing, movement of the config, and movement of the mock objects are all still outstanding.

**Comment by jyellick** [comment]
@vukolic could you look for any potential problems

**Fall Behind State transfer flow**
1. The replica tracks above-watermark checkpoints by replicaID.  If f+1 checkpoints above the high watermark are observed, then this replica is flagged as out of date, and the watermarks are moved, such that the low watermark corresponds to the lowest sequence number of those f+1 checkpoints.
2. The replica participates in ordering normally, but does not execute requests, when it witnesses a weak certificate for a checkpoint, it passes the block hash and number into the state transfer, which synchronizes the state and blockchain to that block in the background.  Note that witnessing a weak certificate for a checkpoint requires that the sequence numbers be within the watermarks, this ensures that the state sync is not told to sync to a block number which is too old.
3.  The replica continues to track above-watermark checkpoints in case the state transfer cannot complete before the network has moved on.  It continues to move its watermarks, and supply weak certificate block and block hash information until the state transfer completes.
4.  Eventually, the state transfer completes, the replica retrieves the sequence number of the current state from the blockchain, marks itself as up to date, and executes any outstanding transactions starting from the next sequence number.  State transfer is complete.

**Comment by jyellick** [comment]
Just as a quick update, this issue is very nearly closed.  It is code complete, with good mock test coverage.  The pending item is live network testing.

**Comment by kchristidis** [comment]
Created PR #543 to push the code in, but let's not close this yet until we have some live network testing done (as @jyellick [indicated above](https://github.com/openblockchain/obc-peer/issues/291#issuecomment-173998123)), and @corecode has integrated it into Sieve.

**Comment by vukolic** [comment]
@jyellick Assuming weak certificate = f+1 same checkpoint messages for the same checkpoint number and same hash - this seems fine - it would be wrong to sync to the f+1st highest checkpoint number but I see you use this scheme just to move watermarks which seems fine. 

That said, I am not sure of the liveness of the scheme. One would need to prevent state transfer from failing all the time. This could happen, for example if, after receiving f+1 matching checkpoint messages, the replica catching up (call it lateReplica) always asks the very same replica (say that one with the lowest ID) for the actual state transfer but this replica, being Byzantine, refuses to give the state to lateReplica, therefore repeatedly failing the state transfer. Some randomization in selection of a state transfer peer or some other approach (asking f+1 replicas for state transfer in parallel) might help there.

**Comment by jyellick** [comment]
@vukolic Great, that is exactly what I was hoping to hear.  To mitigate the liveliness problem, the replica is already selected randomly.  Additionally, as further information, like additional weak checkpoint certificates are received, the set of replicas to sync from is tweaked, so I think we are okay there.

**Comment by kchristidis** [comment]
State transfer is integrated with PBFT classic and batch. Will create a new issue for the ongoing efforts to integrate it with Sieve.
